In [4]:
!python3 -m pip install -U python-woc pandas matplotlib
# please clear the output of this cell in your notebook before checking it in
%pip install python-woc pandas matplotlib tqdm

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 26.0.1 -> 26.2.1
[notice] To update, run: pip install --upgrade pip
Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.4/6.4 MB 10.8 MB/s eta 0:00:0000:010:01m
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.1/11.1 MB 43.4 MB/s eta 0:00:0000:010:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.9/9.9 MB 44.5 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 80.2/80.2 kB 10.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.4/199.4 kB 21.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.3/8.3 MB 27.2 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.9/16.9 MB 26.9 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 355.2/355.2 kB 28.8 MB/s eta 0:00:00

In [10]:
from woc.remote import WocMapsRemote
from tqdm import tqdm
import pandas as pd
# creates the client
woc = WocMapsRemote( base_url="https://worldofcode.org/api/")
# if you got an API key
# woc = WocMapsRemote( base_url="https://worldofcode.org/api/", api_key="woc-XXXXXX-YYYYYY" )

# Now for each of the ten projects you were assigned get commits, e.g.
projects = [
    "mobvoi_wenet",
    "pbs-assess_gfsynopsis",
    "pmlmodelling_ersem",
    "vearch_vearch",
    "hannorein_rebound",
    "soedinglab_mmseqs2",
    "hpddm_hpddm",
    "stephenslab_mr.mash.alpha",
    "gem_oq-engine",
    "sixin-zh_kymatio_wph",
]
list_df_commits = []
for prj in projects:
  # Convert GitHub repository names to WoC V2412 project names
  prj = prj.lower().replace('/','_',2)
  commits = woc.get_values('p2c', prj)
  df = pd.DataFrame(commits, columns=["sha1"])
  df['project'] = prj
  list_df_commits.append(df)
df = pd.concat(list_df_commits)
#perhaps save the list (if no errors) so you do not need to retrieve them again
df.to_csv('df_commits.csv')
df.head(1)

,sha1,project
0,00068789555dbe4c6aff671288cb40ead8bf9105,mobvoi_wenet


In [28]:
# If the number of commits its not very large, you can try to get them all at the same time,
# The max batch size is 10
import time
# let us first split df['sha1'] in chunks

# chunks = [df['sha1'][x:x+40] for x in range(0, len(df), 40)]
commit_data = []

done = {item["commit"] for item in commit_data}
remaining = df.loc[~df["sha1"].isin(done), "sha1"]

chunks = [
    remaining.iloc[x:x+40]
    for x in range(0, len(remaining), 40)
]

for chunk in tqdm(chunks): # iterate over the commits
  # res, err = woc.show_content_many('commit',chunk.to_list())

  # commit.tch returns the same data but faster
  res, err = woc.get_values_many('commit.tch',chunk.to_list())
  res = {k: v[0] for k, v in res.items()}  # this conversion is necessary because of the internal implementation

  # to walk around the rate limit
  time.sleep(4)

  if err: # check for errors
    print('Got Errors', err)

  for commit_sha, commit in res.items():
    # flatten commit objects
    commit_data.append({
          'commit': commit_sha,
          'tree': commit[0],
          'parent': list(commit[1]),
          'author': commit[2][0],
          'author_time': int(commit[2][1]),
          'author_tz': commit[2][2],
          'committer': commit[3][0],
          'committer_time': int(commit[3][1]),
          'committer_tz': commit[3][2],
          'message': commit[4],
    })

df_commit_data = pd.DataFrame(commit_data)
df_commit_data = df_commit_data.merge(
    df[["sha1", "project"]], left_on="commit", right_on="sha1"
)
df_commit_data.to_csv('df_commit_data.csv', index=False)
df_commit_data.head(2)

  5%|▌         | 297/5427 [20:51<5:51:35,  4.11s/it]

Got Errors {'0a1b5bc7422715af441f6f3f825fddb4d078040c': 'Key 0a1b5bc7422715af441f6f3f825fddb4d078040c not found in /da5_fast/All.sha1c/commit_10.tch'}


  8%|▊         | 436/5427 [30:23<5:41:08,  4.10s/it]

Got Errors {'e302bf378a7c58ff716bcbd8b947d7e09a4d1696': 'Key e302bf378a7c58ff716bcbd8b947d7e09a4d1696 not found in /da5_fast/All.sha1c/commit_99.tch'}


  9%|▊         | 468/5427 [32:34<5:39:08,  4.10s/it]

Got Errors {'127786fe91227d444cf26d266be6df7d1fa52bb7': 'Key 127786fe91227d444cf26d266be6df7d1fa52bb7 not found in /da5_fast/All.sha1c/commit_18.tch'}


  9%|▉         | 482/5427 [33:32<5:38:35,  4.11s/it]

Got Errors {'2613ac50074d634ee4ad36437c52baa3eaba1f20': 'Key 2613ac50074d634ee4ad36437c52baa3eaba1f20 not found in /da5_fast/All.sha1c/commit_38.tch'}


 10%|▉         | 535/5427 [37:10<5:37:10,  4.14s/it]

Got Errors {'73d21379b9a3666f5f54ad6928890cc8db894c23': 'Key 73d21379b9a3666f5f54ad6928890cc8db894c23 not found in /da5_fast/All.sha1c/commit_115.tch'}


 10%|▉         | 542/5427 [37:38<5:35:42,  4.12s/it]

Got Errors {'7e17407ed78b697a1ed05804435c6c68d4bfc4fc': 'Key 7e17407ed78b697a1ed05804435c6c68d4bfc4fc not found in /da5_fast/All.sha1c/commit_126.tch'}


 10%|█         | 550/5427 [38:11<5:35:00,  4.12s/it]

Got Errors {'8b6a3679be7281c67cd818664cff8cbbb261a649': 'Key 8b6a3679be7281c67cd818664cff8cbbb261a649 not found in /da5_fast/All.sha1c/commit_11.tch'}


 12%|█▏        | 638/5427 [44:12<5:30:39,  4.14s/it]

Got Errors {'77dc131564b02f129cf618a80eaaabcbb402355a': 'Key 77dc131564b02f129cf618a80eaaabcbb402355a not found in /da5_fast/All.sha1c/commit_119.tch'}


 13%|█▎        | 696/5427 [48:11<5:23:53,  4.11s/it]

Got Errors {'0179f165665f4d14c207c2111da919e099d28cd3': 'Key 0179f165665f4d14c207c2111da919e099d28cd3 not found in /da5_fast/All.sha1c/commit_1.tch'}


 13%|█▎        | 721/5427 [49:54<5:21:54,  4.10s/it]

Got Errors {'02cfb9e61f1524fccd83faf0b86efcad695f0f18': 'Key 02cfb9e61f1524fccd83faf0b86efcad695f0f18 not found in /da5_fast/All.sha1c/commit_2.tch'}


 19%|█▉        | 1048/5427 [1:12:18<4:59:53,  4.11s/it]

Got Errors {'14a4143682262d104a249ba60915f6a6ca0688b4': 'Key 14a4143682262d104a249ba60915f6a6ca0688b4 not found in /da5_fast/All.sha1c/commit_20.tch'}


 20%|██        | 1109/5427 [1:16:29<4:55:11,  4.10s/it]

Got Errors {'17ece029d2a5ada9b68621d368fd101943027637': 'Key 17ece029d2a5ada9b68621d368fd101943027637 not found in /da5_fast/All.sha1c/commit_23.tch'}


 21%|██        | 1132/5427 [1:18:04<4:55:14,  4.12s/it]

Got Errors {'190a6b593ea33921f82fd3b9941ee7cd38478a44': 'Key 190a6b593ea33921f82fd3b9941ee7cd38478a44 not found in /da5_fast/All.sha1c/commit_25.tch'}


 21%|██        | 1148/5427 [1:19:09<4:52:38,  4.10s/it]

Got Errors {'19e5e3b0ca04d8b9e58f462a52ea1361edd6bd1e': 'Key 19e5e3b0ca04d8b9e58f462a52ea1361edd6bd1e not found in /da5_fast/All.sha1c/commit_25.tch'}


 22%|██▏       | 1217/5427 [1:23:53<4:48:39,  4.11s/it]

Got Errors {'1d90533b93e7304cba4ac20f33cb0c6279e17935': 'Key 1d90533b93e7304cba4ac20f33cb0c6279e17935 not found in /da5_fast/All.sha1c/commit_29.tch'}


 27%|██▋       | 1444/5427 [1:39:26<4:32:40,  4.11s/it]

Got Errors {'29bd81b16d15ba546216f97a424a1ac02172fbe5': 'Key 29bd81b16d15ba546216f97a424a1ac02172fbe5 not found in /da5_fast/All.sha1c/commit_41.tch'}


 30%|██▉       | 1627/5427 [1:51:59<4:20:24,  4.11s/it]

Got Errors {'33936bbe0e2d3c0f1180559fcda831c87b5cdf4f': 'Key 33936bbe0e2d3c0f1180559fcda831c87b5cdf4f not found in /da5_fast/All.sha1c/commit_51.tch'}


 31%|███       | 1685/5427 [1:55:57<4:15:56,  4.10s/it]

Got Errors {'36b2985606efde558bcad7d33353cd43b54e3757': 'Key 36b2985606efde558bcad7d33353cd43b54e3757 not found in /da5_fast/All.sha1c/commit_54.tch'}


 32%|███▏      | 1729/5427 [1:58:58<4:12:56,  4.10s/it]

Got Errors {'38f9522803fa1f89bd01fccbef240705ed4c1f1f': 'Key 38f9522803fa1f89bd01fccbef240705ed4c1f1f not found in /da5_fast/All.sha1c/commit_56.tch'}


 33%|███▎      | 1801/5427 [2:03:54<4:08:12,  4.11s/it]

Got Errors {'3ce93c6b6b038ae87b21d3709d40bd63e58403e1': 'Key 3ce93c6b6b038ae87b21d3709d40bd63e58403e1 not found in /da5_fast/All.sha1c/commit_60.tch'}


 34%|███▍      | 1865/5427 [2:08:18<4:04:40,  4.12s/it]

Got Errors {'405c64ee6495abab12fb32271df71fa02ce53531': 'Key 405c64ee6495abab12fb32271df71fa02ce53531 not found in /da5_fast/All.sha1c/commit_64.tch'}


 42%|████▏     | 2290/5427 [2:37:25<3:34:44,  4.11s/it]

Got Errors {'5783bcd8ad54d6a35e3cf26785b3ccf8e1baadac': 'Key 5783bcd8ad54d6a35e3cf26785b3ccf8e1baadac not found in /da5_fast/All.sha1c/commit_87.tch'}


 44%|████▍     | 2392/5427 [2:44:24<3:28:18,  4.12s/it]

Got Errors {'5d276ae6fbebc400a49d4c90346fc658fd6a4504': 'Key 5d276ae6fbebc400a49d4c90346fc658fd6a4504 not found in /da5_fast/All.sha1c/commit_93.tch'}


 44%|████▍     | 2401/5427 [2:45:01<3:26:51,  4.10s/it]

Got Errors {'5d9fed5ebf24d0affc74210a2a1c25d4f98aa8a1': 'Key 5d9fed5ebf24d0affc74210a2a1c25d4f98aa8a1 not found in /da5_fast/All.sha1c/commit_93.tch'}


 47%|████▋     | 2530/5427 [2:53:51<3:18:15,  4.11s/it]

Got Errors {'648864d6262e79f9ce6d64714b507326ad43573a': 'Key 648864d6262e79f9ce6d64714b507326ad43573a not found in /da5_fast/All.sha1c/commit_100.tch'}


 51%|█████▏    | 2783/5427 [3:11:10<3:01:55,  4.13s/it]

Got Errors {'721a1f268b165e6a985a1c164d1ee308621d6096': 'Key 721a1f268b165e6a985a1c164d1ee308621d6096 not found in /da5_fast/All.sha1c/commit_114.tch'}


 54%|█████▍    | 2922/5427 [3:20:41<2:51:08,  4.10s/it]

Got Errors {'7998cfcbafa32f83a060f87b45cacfbd43480c85': 'Key 7998cfcbafa32f83a060f87b45cacfbd43480c85 not found in /da5_fast/All.sha1c/commit_121.tch'}


 55%|█████▌    | 2988/5427 [3:25:13<2:46:57,  4.11s/it]

Got Errors {'7d0cd19bed24fa366c0d0530335959a4c8c45c7e': 'Key 7d0cd19bed24fa366c0d0530335959a4c8c45c7e not found in /da5_fast/All.sha1c/commit_125.tch'}


 58%|█████▊    | 3135/5427 [3:35:16<2:36:31,  4.10s/it]

Got Errors {'85078aafb8aca35b2105878277fb98f8df7bde43': 'Key 85078aafb8aca35b2105878277fb98f8df7bde43 not found in /da5_fast/All.sha1c/commit_5.tch'}


 63%|██████▎   | 3444/5427 [3:56:24<2:15:27,  4.10s/it]

Got Errors {'95d05faa05718dfa35117bab82d230137f6b365e': 'Key 95d05faa05718dfa35117bab82d230137f6b365e not found in /da5_fast/All.sha1c/commit_21.tch'}


 64%|██████▎   | 3449/5427 [3:56:44<2:15:15,  4.10s/it]

Got Errors {'961895de214fad1dcabc6291d61ac18574d6ba2e': 'Key 961895de214fad1dcabc6291d61ac18574d6ba2e not found in /da5_fast/All.sha1c/commit_22.tch'}


 65%|██████▍   | 3518/5427 [4:01:27<2:10:56,  4.12s/it]

Got Errors {'99e6209bbbdf6277e5b268c6373da6c319573928': 'Key 99e6209bbbdf6277e5b268c6373da6c319573928 not found in /da5_fast/All.sha1c/commit_25.tch'}


 66%|██████▌   | 3594/5427 [4:06:39<2:05:24,  4.10s/it]

Got Errors {'9e137ebfc2a26779f5eaf07188f74f85b262e7a2': 'Key 9e137ebfc2a26779f5eaf07188f74f85b262e7a2 not found in /da5_fast/All.sha1c/commit_30.tch'}


 68%|██████▊   | 3709/5427 [4:14:31<1:57:30,  4.10s/it]

Got Errors {'a42e236bddd0b546756e65bd0082d8f92df538b1': 'Key a42e236bddd0b546756e65bd0082d8f92df538b1 not found in /da5_fast/All.sha1c/commit_36.tch'}


 76%|███████▌  | 4106/5427 [4:41:40<1:30:14,  4.10s/it]

Got Errors {'b9288ca593c81230ae6cf2054bb24eebe52445fc': 'Key b9288ca593c81230ae6cf2054bb24eebe52445fc not found in /da5_fast/All.sha1c/commit_57.tch'}


 80%|████████  | 4345/5427 [4:58:01<1:14:03,  4.11s/it]

Got Errors {'c5fe0cfecb55fff1e2db22bc88acb464d48f6ed2': 'Key c5fe0cfecb55fff1e2db22bc88acb464d48f6ed2 not found in /da5_fast/All.sha1c/commit_69.tch'}


 83%|████████▎ | 4525/5427 [5:10:20<1:01:36,  4.10s/it]

Got Errors {'cfe4fa628308a0d88d52b8960dfd0647983e06f8': 'Key cfe4fa628308a0d88d52b8960dfd0647983e06f8 not found in /da5_fast/All.sha1c/commit_79.tch'}


 84%|████████▍ | 4570/5427 [5:13:24<58:43,  4.11s/it]  

Got Errors {'d23569243e5bf9dd89288f54eda750ae9a1c9298': 'Key d23569243e5bf9dd89288f54eda750ae9a1c9298 not found in /da5_fast/All.sha1c/commit_82.tch'}


 86%|████████▌ | 4654/5427 [5:19:09<52:57,  4.11s/it]

Got Errors {'d6bfcf26c2cc891e6f847cf7baf71c9fe05464dc': 'Key d6bfcf26c2cc891e6f847cf7baf71c9fe05464dc not found in /da5_fast/All.sha1c/commit_86.tch'}


 87%|████████▋ | 4721/5427 [5:23:44<48:10,  4.09s/it]

Got Errors {'da6115ec86d65897bd4fb51165cb3fb3f2b8daf9': 'Key da6115ec86d65897bd4fb51165cb3fb3f2b8daf9 not found in /da5_fast/All.sha1c/commit_90.tch'}


 90%|█████████ | 4905/5427 [5:36:19<35:44,  4.11s/it]

Got Errors {'e46e4e930398a55397bfedbd5099e0d058e38eaf': 'Key e46e4e930398a55397bfedbd5099e0d058e38eaf not found in /da5_fast/All.sha1c/commit_100.tch'}


 91%|█████████ | 4925/5427 [5:37:41<34:22,  4.11s/it]

Got Errors {'e58d6e39872f43e3deacf2949e796468cabd8554': 'Key e58d6e39872f43e3deacf2949e796468cabd8554 not found in /da5_fast/All.sha1c/commit_101.tch'}


 95%|█████████▍| 5153/5427 [5:53:17<18:41,  4.09s/it]

Got Errors {'f1fad75b4ace513580d10c655340aed5c94e5cfd': 'Key f1fad75b4ace513580d10c655340aed5c94e5cfd not found in /da5_fast/All.sha1c/commit_113.tch'}


 98%|█████████▊| 5331/5427 [6:05:27<06:33,  4.10s/it]

Got Errors {'fba2df9ec88cdeecda163012eb6b8f6adcd4dc30': 'Key fba2df9ec88cdeecda163012eb6b8f6adcd4dc30 not found in /da5_fast/All.sha1c/commit_123.tch'}


100%|██████████| 5427/5427 [6:12:01<00:00,  4.11s/it]


,commit,tree,parent,author,author_time,author_tz,committer,committer_time,committer_tz,message,sha1,project
0,00068789555dbe4c6aff671288cb40ead8bf9105,2331e07c0d80174bdbd36457128e47d9f79edcd6,[87886a897ec6c7da0fee38cf8ba266a52c76dc65],Binbin Zhang <binbzha@qq.com>,1651152697,+0800,Binbin Zhang <binbzha@qq.com>,1651152697,+0800,[binding]: add python binding\n,00068789555dbe4c6aff671288cb40ead8bf9105,mobvoi_wenet
1,000c7af5c36afafd3fa3ae4599af237e902c67bb,634e826cace15c002e2d0820a9a961af4e8111b5,[980507ecaba296a363ddb9a1cd8f93be6b2e8b50],Mddct <hamddct@gmail.com>,1697436927,+0800,Mddct <hamddct@gmail.com>,1697785167,+0800,[cli/paraformer] ali-paraformer load and inf...,000c7af5c36afafd3fa3ae4599af237e902c67bb,mobvoi_wenet


In [23]:
import httpx

try:
    response = httpx.get(
        "https://worldofcode.org/api/",
        timeout=20
    )
    print("Connected. HTTP status:", response.status_code)
except httpx.RequestError as error:
    print(type(error).__name__, str(error))
    print("Underlying cause:", repr(error.__cause__))

Connected. HTTP status: 404


In [7]:
# what does the first record looks like?
# commit  parents
v = commit_data[0]
# commit tree, [parent{s}], [author, unixtime, tz ], [ committer, unixtime, tz ], Commit_message ]
print(v)

{'commit': '00144f85329256bf24124729a0554f76fda95877', 'tree': '116d1fe9034b233f623863ccaac7f9b597013d68', 'parent': ['f254d1c0a5e9e80b9f7e8d3b4709b9e93d52f618'], 'author': 'mmorise <mmorise@yamanashi.ac.jp>', 'author_time': 1492872352, 'author_tz': '+0900', 'committer': 'mmorise <mmorise@yamanashi.ac.jp>', 'committer_time': 1492872352, 'committer_tz': '+0900', 'message': 'DC component is removed\n'}


In [8]:
# Now that the data is retrieved, save it and commit to your GH fork
# First, we need to flatten info in order to export as csv
# our dataframe will have columns project,'commit, author, time, message'
dfinf = pd.DataFrame(columns=['project', 'commit', 'author', 'time', 'message'])
for k in commit_data:
  row = pd.Series({'project':prj, 'commit': k['commit'], 'author': k['author'], 'time': k['author_time'], 'message':k['message']})
  dfinf = pd.concat([dfinf, row.to_frame().T ], ignore_index=True)


In [9]:
# check if it has the right content
dfinf.head(1)

,project,commit,author,time,message
0,mmorise_world,00144f85329256bf24124729a0554f76fda95877,mmorise <mmorise@yamanashi.ac.jp>,1492872352,DC component is removed\n


In [ ]:
#mode a means append, so you have all your projects in the same file
yournetid='your_net_id_here'
dfinf.to_csv(yournetid+'_project_summary.csv', index=False,sep=';', mode='a', header=False)

# Make sure you check in to your fork not just the notebook but also the csv files!!!

# Don't forget to add requested data from github and this notebook
### For each of the 10 projects go to their github repo and get the number of stars, number of forks, and the last commit date
### Report (in your notebook) the number of commits, the number of authors, and max and min time for each project based on WoC commits and also add the info you obtained from github

## GitHub project information

Current observations; each row records its collection time in UTC. Last commit uses the default branch's latest listed commit and its committer timestamp. This is not the historical September 30, 2025 status assessment.

| Project | Stars | Forks | Last commit (UTC) | Branch | Checked (UTC) | Repository | Status |
|---|---:|---:|---|---|---|---|---|
| mobvoi_wenet | 5243 | 1190 | 2026-09-07T07:21:08Z | main | 2026-10-08T03:42:48.905413+00:00 | https://github.com/wenet-e2e/wenet | ok |
| pbs-assess_gfsynopsis | 17 | 1 | 2026-09-19T23:19:56Z | master | 2026-10-08T03:42:51.508621+00:00 | https://github.com/pbs-assess/gfsynopsis | ok |
| pmlmodelling_ersem | 30 | 14 | 2026-08-18T13:51:19Z | master | 2026-10-08T03:42:54.094392+00:00 | https://github.com/pmlmodelling/ersem | ok |
| vearch_vearch | 2331 | 365 | 2026-07-27T05:29:53Z | master | 2026-10-08T03:42:56.587535+00:00 | https://github.com/vearch/vearch | ok |
| hannorein_rebound | 1138 | 270 | 2026-10-06T12:13:52Z | main | 2026-10-08T03:42:59.146263+00:00 | https://github.com/hannorein/rebound | ok |
| soedinglab_mmseqs2 | 2148 | 303 | 2026-09-30T04:36:10Z | master | 2026-10-08T03:43:01.645441+00:00 | https://github.com/soedinglab/MMseqs2 | ok |
| hpddm_hpddm | 160 | 37 | 2026-09-26T19:46:30Z | main | 2026-10-08T03:43:04.196691+00:00 | https://github.com/hpddm/hpddm | ok |
| stephenslab_mr.mash.alpha | 7 | 5 | 2026-02-16T16:00:01Z | master | 2026-10-08T03:43:07.010691+00:00 | https://github.com/stephenslab/mr.mashr | ok |
| gem_oq-engine | 451 | 325 | 2026-10-08T02:09:51Z | master | 2026-10-08T03:43:09.478557+00:00 | https://github.com/gem/oq-engine | ok |
| sixin-zh_kymatio_wph | 0 | 0 | 2025-04-10T14:37:11Z | master | 2026-10-08T03:43:12.093200+00:00 | https://github.com/sixin-zh/kymatio_wph | ok |
